### ISSUE TITLE
Auth Code Browser flow throws 403 when login page left idle for >30 mins

### ISSUE THREAD

**Issue by bondugula** [cloudfoundry/uaa#445]
When trying to do the authcode browser flow through uaac, once we enter the client secret , the browser redirects to login page. When we leave the login page idle for 30 mins, and then try to login. The first time we enter the username and password, it fails with following error "Invalid login attempt, request does not meet our security standards, please try again."

Please find the attached screenshot.
<img width="697" alt="screen shot 2016-09-20 at 5 00 38 pm" src="https://cloud.githubusercontent.com/assets/13593614/18693018/60bf9e72-7f53-11e6-861a-4db56eca07d2.png">

Steps to reproduce:
1. git clone open source release version of UAA 3.6.0
2. ./gradlew run ( start uaa locally)
3. Once UAA is started, target the localhost uaa , login with admin client, then do the authcode browser flow
uaac target http://localhost:8080/uaa
uaac token authcode get -c oauth_showcase_authorization_code
This prompts to enter client secret, once we enter the secret, it redirects to the UAA login page and leave the page idle for 30mins and then try to enter the username "testbootuser" and the password. This throws the error. But if we do the same flow without leaving the login page idle for 30 mins, it is successful.

**Comment by amiri-ge** [comment]
I have some more information to add to this issue. After waiting more than 30 minutes and logging in, the login fails even though the credentials are correct. Furthermore, if I try to login in again, while the login does work the second time, it does not redirect me to the original location that was specified in the OAuth authorization code grant redirect_uri query string parameter. It is is as if some internal session expires, and when it does UAA forgets what the redirect_uri of the original request was.

To reproduce:
1. git clone open source release version of UAA 3.6.0
2. ./gradlew run
3. uaac token authcode get -c oauth_showcase_authorization_code -s secret
4. After the login page loads in your browser wait for more than 30 minutes
5. After more than 30 minutes login using the standard marissa/koala username/password
6. Observe the initial failure
7. Try logging in again with marissa/koala
8. Note that, though the login is successful, the Location header of the 302 response is to /uaa/ instead of the request_uri specified by the uaac initiated OAuth authorization code grant request
9. UAA is left hanging and everyone is so sad...

**Comment by john-k-ge** [comment]
I looked at this for quite a bit for another user, and I don't need to wait 30 mins; a little over 5 will do.  Here's my two cents:

When the user lands on the UAA logon page, a cross-site forgery cookie (X-Uaa-Csrf) is set.  This expires in 5 minutes.  If the page is left idle for 5 mins, any attempt to logon (good or bad creds) results in the "Invalid login attempt, request does not meet our security standards...". You can confirm this.  This appears to be standard UAA behavior, as evidenced by this comment in https://github.com/cloudfoundry/uaa/issues/348 :

"...The UAA login page is protected against cross site request forgery. When the login page is requested by the browser, a session is created and a so called CSRF token is created. Please check if the session expired and reload the UAA login page. Usually this resolves the issue as the CSRF token and the session are renewed...."

I confirmed that the cookie TTL is 300 seconds--5 mins in CookieBasedCsrfTokenRepository.java  This would explain why subsequent logons work fine:  after the initial failure, users try again right away.

**Comment by fhanik** [comment]
If you need that timeout to be longer, on a fork you could set the timeout longer

    <bean id="loginCookieCsrfRepository" class="org.cloudfoundry.identity.uaa.security.web.CookieBasedCsrfTokenRepository">
        <property name="secure" value="${require_https:false}" />
        <property name="cookieMaxAge" value="<time in seconds>"/>
    </bean>

https://github.com/cloudfoundry/uaa/blob/cded6164a3b90e791688a954069aea3cfde59b69/server/src/main/java/org/cloudfoundry/identity/uaa/security/web/CookieBasedCsrfTokenRepository.java#L31-L31

A simple PR, and that value can be exposed as a Yaml configuration parameter

Not using any CSRF protection would cause this to be a security issue. And you could set timeout to a year if that would solve your issue.

**Comment by amiri-ge** [comment]
@fhanik @john-k-ge I think you're both missing an important aspect of the problem. It's not just that the CSRF token expires, that's actually not the bigger issue here. The bigger issue is that when the CSRF token expires, the original OAuth context is completely lost and the user finds themselves staring at a login page with the presumption that if they try again and login in successfully they will login in to the application that redirected them. Unfortunately, that's not what happens. When UAA loses the original OAuth context, the subsequent login page is setup as a login _into UAA itself_.

In summary, the current behavior is not a graceful way to fail on CSRF token expiration for logins that originate from an OAuth flow. A better UX would be to redirect the user back to the application that initiated the OAuth authorization code grant request with an OAuth error.

In other words the current flow breaks OAuth and it really shouldn't. It should handle this as it would any other OAuth error.

**Comment by fhanik** [comment]
Rephrasing this issue to storing `/oauth/authorize` state on the client side.
Initial implementation will be simple flow only, will not cover forgot password and other intermittent flows
